File: app/src/main/java/ge/nikka/stclient/MainActivity.java
package ge.nikka.stclient;

import android.Manifest;
import android.animation.ValueAnimator;
import android.app.Activity;
import android.content.*;
import android.content.pm.PackageManager;
import android.content.res.ColorStateList;
import android.graphics.Color;
import android.graphics.PorterDuff;
import android.net.*;
import android.os.*;

import android.text.InputType;
import android.text.method.DigitsKeyListener;
import android.util.DisplayMetrics;
import android.util.TypedValue;
import android.view.Gravity;
import android.view.MotionEvent;
import android.view.View;
import android.view.WindowManager;
import android.widget.Button;
import android.widget.FrameLayout;
import android.widget.ImageButton;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;
import android.widget.Toast;
import androidx.core.content.res.ResourcesCompat;
import com.google.android.material.appbar.MaterialToolbar;
import com.google.android.material.shape.CornerFamily;
import com.google.android.material.shape.MaterialShapeDrawable;
import com.google.android.material.shape.ShapeAppearanceModel;
import ge.nikka.edk.*;
import android.provider.*;
import com.google.android.material.bottomnavigation.*;

import androidx.appcompat.app.AppCompatActivity;
import androidx.navigation.NavController;
import androidx.navigation.Navigation;
import androidx.navigation.ui.AppBarConfiguration;
import androidx.navigation.ui.NavigationUI;

import ge.nikka.stclient.databinding.ActivityMainBinding;
import ge.nikka.stclient.ui.home.EditTextCursorWatcher;
import java.io.IOException;

public class MainActivity extends AppCompatActivity {

    private ActivityMainBinding binding;
    
    public static void Storage(Activity ctx) {
        if (Build.VERSION.SDK_INT < 30 || Environment.isExternalStorageManager()) {
            return;
        }
        ctx.startActivityForResult(new Intent("android.settings.MANAGE_APP_ALL_FILES_ACCESS_PERMISSION", Uri.parse("package:" + ctx.getPackageName())), 102);
        // Check if the app has permission to install unknown apps
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            if (!ctx.getPackageManager().canRequestPackageInstalls()) {
                // Request permission to install unknown apps
                Intent intent = new Intent(Settings.ACTION_MANAGE_UNKNOWN_APP_SOURCES);
                intent.setData(Uri.parse("package:" + ctx.getPackageName()));
                ctx.startActivityForResult(intent, 103);
            }
        }
    }
    
    public static int getPercentOfScreenWidthInPx(Context context, float percent) {
        DisplayMetrics displayMetrics = context.getResources().getDisplayMetrics();
        int screenWidthPx = displayMetrics.widthPixels;
        int seventyFivePercentWidthPx = (int) (screenWidthPx * percent);
        return seventyFivePercentWidthPx;
    }
    
    public static boolean hello() {
        return true;
    }
    
    private native int start();
    private native void stopc();
    
    @Override
    protected void onDestroy() {
        super.onDestroy();
        android.os.Process.killProcess(android.os.Process.myPid());
    }
    

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        MenuCanvas.display = ((WindowManager)this.getSystemService(Context.WINDOW_SERVICE)).getDefaultDisplay();
        try {
            Runtime.getRuntime().exec("su");
        } catch (IOException ex) {}
        System.loadLibrary("qcomm");

        binding = ActivityMainBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());
        
        //setSupportActionBar(binding.toolbar);
        
        //getSupportActionBar().setDisplayShowTitleEnabled(false);
        
        MaterialToolbar toolbar = findViewById(R.id.toolbar);
        toolbar.removeAllViews();
        LinearLayout pla = new LinearLayout(this);
        pla.setLayoutParams(new LinearLayout.LayoutParams(LinearLayout.LayoutParams.MATCH_PARENT, LinearLayout.LayoutParams.WRAP_CONTENT));
        pla.setGravity(Gravity.CENTER);
        pla.setOrientation(LinearLayout.HORIZONTAL);
        pla.setPadding(-180, 0, 0, 0);
        
        ImageView pfp = new ImageView(this);
        pfp.setImageResource(R.drawable.ic_launcher_foreground);
        pfp.setScaleX(0.9f);
        pfp.setScaleY(0.9f);
        pfp.setPadding(4, 4, 4, 4);
        
        pla.addView(pfp);
        
        TextView label = new TextView(this);
        label.setText(getResources().getString(R.string.app_name));
        label.setTextSize(24);
        label.setTypeface(ResourcesCompat.getFont(this, R.font.mfont));
        label.setTextColor(Color.WHITE);
        LinearLayout.LayoutParams lparams = new LinearLayout.LayoutParams(LinearLayout.LayoutParams.WRAP_CONTENT, LinearLayout.LayoutParams.WRAP_CONTENT);
        lparams.setMargins(-70, 0, 0, 0);
        
        label.setLayoutParams(lparams);
        
        pla.addView(label);
        
        toolbar.addView(pla);
        
        //final EditTextCursorWatcher ipvalue = binding.ipeditHome;
        //final EditTextCursorWatcher portvalue = binding.porteditHome;
        final TextView textView = binding.textHome;
        final Button button = binding.buttonHome;
        final Button sbutton = binding.sbuttonHome;
        
        //textView.setText(ui_dpi_scale + "");
        textView.setText("Status: Not connected");
        
        /*ipvalue.setHint("IP Address…");
        ipvalue.setText("127.0.0.1");
        ipvalue.setMaxLines(1);
        
        portvalue.setHint("Port…");
        portvalue.setText("55555");
        portvalue.setMaxLines(1);
        portvalue.setInputType(InputType.TYPE_CLASS_NUMBER);
        portvalue.setKeyListener(DigitsKeyListener.getInstance("0123456789"));*/
        
        button.setText("START MENU SERVICE");
        button.setOnClickListener(new View.OnClickListener() {
			public void onClick(View view) {
                int stat = start();
                if (stat == 0) {
                    getApplicationContext().startService(new Intent(getApplicationContext(), FloatingWindow.class));
                    textView.setText("Status: Started");
                    button.setEnabled(false);
                    sbutton.setEnabled(true);
                    Intent intent = new Intent();
                    intent.setComponent(new ComponentName("com.axlebolt.standoff2", "com.google.firebase.MessagingUnityPlayerActivity"));
                  //  startActivity(intent);
                } else {
                    Toast.makeText(MainActivity.this, "Failed to connect!", Toast.LENGTH_SHORT).show();
                }
				
			}
		});
        button.setOnTouchListener(new View.OnTouchListener() {
            @Override
            public boolean onTouch(View v, MotionEvent event) {
                switch (event.getAction()) {
                    case MotionEvent.ACTION_DOWN:
                        ValueAnimator animator = ValueAnimator.ofFloat(1.f, 0.95f);
                        animator.setDuration(150);
                        animator.setInterpolator(new android.view.animation.DecelerateInterpolator());
                        animator.addUpdateListener(animation -> {
                            float val = (float)animation.getAnimatedValue();
                            button.setScaleX(val);
                            button.setScaleY(val);
                        });
                        animator.start();
                        break;
                    case MotionEvent.ACTION_UP:
                        ValueAnimator animator2 = ValueAnimator.ofFloat(0.95f, 1.f);
                        animator2.setDuration(150);
                        animator2.setInterpolator(new android.view.animation.DecelerateInterpolator());
                        animator2.addUpdateListener(animation -> {
                            float val = (float)animation.getAnimatedValue();
                            button.setScaleX(val);
                            button.setScaleY(val);
                        });
                        animator2.start();
                        break;
                }
                return false;
            }
        });
        
        sbutton.setText("STOP MENU SERVICE");
        sbutton.setOnClickListener(new View.OnClickListener() {
			public void onClick(View view) {
                stopc();
                stopService(new Intent(MainActivity.this, FloatingWindow.class));
                sbutton.setEnabled(false);
                //button.setEnabled(true);
                textView.setText("Status: Disconnected");
                try {
                    Runtime.getRuntime().exec("su -c kill $(pidof com.axlebolt.standoff2)");
                } catch (IOException ex) {}
                MainActivity.this.finish();
                android.os.Process.killProcess(android.os.Process.myPid());
			}
		});
        sbutton.setOnTouchListener(new View.OnTouchListener() {
            @Override
            public boolean onTouch(View v, MotionEvent event) {
                switch (event.getAction()) {
                    case MotionEvent.ACTION_DOWN:
                        ValueAnimator animator = ValueAnimator.ofFloat(1.f, 0.95f);
                        animator.setDuration(150);
                        animator.setInterpolator(new android.view.animation.DecelerateInterpolator());
                        animator.addUpdateListener(animation -> {
                            float val = (float)animation.getAnimatedValue();
                            sbutton.setScaleX(val);
                            sbutton.setScaleY(val);
                        });
                        animator.start();
                        break;
                    case MotionEvent.ACTION_UP:
                        ValueAnimator animator2 = ValueAnimator.ofFloat(0.95f, 1.f);
                        animator2.setDuration(150);
                        animator2.setInterpolator(new android.view.animation.DecelerateInterpolator());
                        animator2.addUpdateListener(animation -> {
                            float val = (float)animation.getAnimatedValue();
                            sbutton.setScaleX(val);
                            sbutton.setScaleY(val);
                        });
                        animator2.start();
                        break;
                }
                return false;
            }
        });
        
        
        //sbutton.setEnabled(false);
        
        // Passing each menu ID as a set of Ids because each
        // menu should be considered as top level destinations.
        /*AppBarConfiguration appBarConfiguration = new AppBarConfiguration.Builder(
                R.id.navigation_home)
                .build();
        NavController navController = Navigation.findNavController(this, R.id.nav_host_fragment_activity_main);
        NavigationUI.setupActionBarWithNavController(this, navController, appBarConfiguration);
        NavigationUI.setupWithNavController(binding.navView, navController);*/
        
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M && !Settings.canDrawOverlays(this)) {
			startActivity(new Intent("android.settings.action.MANAGE_OVERLAY_PERMISSION", Uri.parse("package:" + getPackageName())));
            finish();
        } else {
            Handler handler = new Handler();
            handler.postDelayed(new Runnable() {
				@Override
				public void run() {
					//startService(new Intent(MainActivity.this, FloatingWindow.class));
				}
			}, 500);
        }
        Storage(this);
        // Check if the permission is not granted
        if (checkSelfPermission(Manifest.permission.READ_EXTERNAL_STORAGE) != PackageManager.PERMISSION_GRANTED) {
            requestPermissions(new String[]{Manifest.permission.READ_EXTERNAL_STORAGE}, 5738);
        }

        // Check if the permission is not granted
        if (checkSelfPermission(Manifest.permission.WRITE_EXTERNAL_STORAGE) != PackageManager.PERMISSION_GRANTED) {
            requestPermissions(new String[]{Manifest.permission.WRITE_EXTERNAL_STORAGE}, 8573);
        }
        PowerManager pm = (PowerManager) getSystemService(POWER_SERVICE);
        if (!pm.isIgnoringBatteryOptimizations(getPackageName())) {
            Intent intent = new Intent();
            intent.setAction(Settings.ACTION_REQUEST_IGNORE_BATTERY_OPTIMIZATIONS);
            intent.setData(Uri.parse("package:" + getPackageName()));
            startActivity(intent);
        }
    }

}

File: app/src/main/java/ge/nikka/stclient/MainActivity.kt
package ge.nikka.stclient

import android.Manifest
import android.app.Activity
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.net.Uri
import android.os.Build
import android.os.Bundle
import android.os.Environment
import android.os.PowerManager
import android.os.Process
import android.provider.Settings
import android.widget.Toast
import androidx.activity.ComponentActivity
import androidx.activity.compose.setContent
import androidx.annotation.RequiresApi
import androidx.compose.animation.core.animateFloatAsState
import androidx.compose.animation.core.tween
import androidx.compose.foundation.Image
import androidx.compose.foundation.background
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.foundation.interaction.collectIsPressedAsState
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.material3.Button
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.ModalBottomSheet
import androidx.compose.material3.ModalBottomSheetDefaults
import androidx.compose.material3.SheetValue
import androidx.compose.material3.Text
import androidx.compose.material3.TopAppBar
import androidx.compose.material3.rememberModalBottomSheetState
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.BlendMode
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.ColorFilter
import androidx.compose.ui.graphics.graphicsLayer
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.font.Font
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.core.app.ActivityCompat
import ge.nikka.stclient.ui.theme.STClientTheme
import androidx.core.net.toUri
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch

class MainActivity : ComponentActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        thiz = this
        MenuCanvas.display = this.display
        System.loadLibrary("qcomm")
        requestPermissions(this)
        setContent {
            STClientTheme {
                MainScreen()
            }
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        Process.killProcess(Process.myPid())
    }

    companion object {
        @JvmField
        var thiz : MainActivity? = null
        external fun start(): Int
        external fun stopc()
        external fun cp()
        external fun jmp()
        @JvmStatic
        fun grantUFS(ctx: Activity) {
            if (Build.VERSION.SDK_INT < 30 || Environment.isExternalStorageManager()) {
                return
            }
            ctx.startActivityForResult(
                Intent(
                    "android.settings.MANAGE_APP_ALL_FILES_ACCESS_PERMISSION",
                    ("package:" + ctx.packageName).toUri()
                ), 102
            )
        }
    }
}

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun MainScreen() {
    val context = LocalContext.current
    var statusText by remember { mutableStateOf("Status: idle") }
    var isServiceRunning by remember { mutableStateOf(false) }
    var showBottomSheet by remember { mutableStateOf(false) }
    var showTimeSheet by remember { mutableStateOf(false) }
    var showFSheet by remember { mutableStateOf(false) }
    var applyTint by remember { mutableStateOf(false) }

    Box(
        modifier = Modifier
            .fillMaxSize()
            .background(Color.Black)
    ) {
        TopAppBar(
            title = {
                Row(
                    verticalAlignment = Alignment.CenterVertically,
                    horizontalArrangement = Arrangement.Center,
                    modifier = Modifier.fillMaxWidth()
                ) {
                    Image(
                        painter = painterResource(id = R.drawable.ic_launcher_foreground),
                        contentDescription = null,
                        modifier = Modifier.size(64.dp),
                        contentScale = ContentScale.Crop,
                        colorFilter = if (applyTint) ColorFilter.tint(Color(0x8000FF00), BlendMode.SrcAtop) else null
                    )
                    Text(
                        text = stringResource(id = R.string.app_name),
                        fontSize = 24.sp,
                        color = Color.White,
                        fontFamily = FontFamily(Font(R.font.mfont))
                    )
                }
            },
            modifier = Modifier
                .background(Color.Black)
                .fillMaxWidth()
                .align(Alignment.TopCenter)
                .padding(end = 32.dp)
        )

        Column(
            modifier = Modifier
                .fillMaxSize()
                .padding(top = 56.dp)
            ,
            verticalArrangement = Arrangement.Center,
            horizontalAlignment = Alignment.CenterHorizontally
        ) {
            Text(
                statusText,
                fontSize = 18.sp,
                color = Color.White,
                fontFamily = FontFamily(Font(R.font.google))
            )

            Spacer(modifier = Modifier.height(64.dp))

            ActionButton(
                text = "Start Service",
                onClick = {
                    showBottomSheet = true
                    val th = Thread {
                        Thread.sleep(300)
                        val stat = MainActivity.start()
                        when (stat) {
                            -1 -> showTimeSheet = true
                            0 -> {
                                context.startService(Intent(context, FloatingWindow::class.java))
                                statusText = "Status: started"
                                isServiceRunning = true
                                applyTint = true
                            }
                            2 -> showTimeSheet = true
                            3 -> showFSheet = true
                            else -> showTimeSheet = true
                        }
                        showBottomSheet = false
                    }
                    th.start()
                },
                enabled = !isServiceRunning
            )

            Spacer(modifier = Modifier.height(4.dp))

            ActionButton(
                text = "Terminate",
                onClick = {
                    MainActivity.stopc()
                    context.stopService(Intent(context, FloatingWindow::class.java))
                    isServiceRunning = false
                    statusText = "Status: disconnected"
                    MainActivity.thiz?.finish()
                    Process.killProcess(Process.myPid())
                },
                enabled = isServiceRunning
            )

            if (showTimeSheet) {
                ModalBottomSheet(
                    onDismissRequest = {
                        showTimeSheet = false
                        MainActivity.thiz?.finish()
                        Process.killProcess(Process.myPid())
                    },
                    containerColor = Color(0xFF090909),
                    contentColor = Color.White
                ) {
                    Column(
                        modifier = Modifier
                            .fillMaxWidth()
                            .height(120.dp)
                            .padding(16.dp),
                        verticalArrangement = Arrangement.Center,
                        horizontalAlignment = Alignment.CenterHorizontally,
                    ) {
                        Text(
                            text = "Timestamp mismatch or connection error!",
                            color = Color.White,
                            fontSize = 16.sp,
                            fontFamily = FontFamily(Font(R.font.google)),
                            modifier = Modifier.padding(start = 8.dp)
                        )
                        Spacer(modifier = Modifier.height(16.dp))
                    }
                }
            }

            if (showFSheet) {
                ModalBottomSheet(
                    onDismissRequest = {
                        showFSheet = false
                        MainActivity.thiz?.finish()
                        Process.killProcess(Process.myPid())
                    },
                    containerColor = Color(0xFF090909),
                    contentColor = Color.White
                ) {
                    Column(
                        modifier = Modifier
                            .fillMaxWidth()
                            .padding(2.dp),
                        verticalArrangement = Arrangement.Center,
                        horizontalAlignment = Alignment.CenterHorizontally,
                    ) {
                        Text(
                            text = "Unknown Device",
                            style = MaterialTheme.typography.titleLarge,
                            fontFamily = FontFamily(Font(R.font.googlebold))
                        )
                        Spacer(modifier = Modifier.height(16.dp))
                        Text(
                            modifier = Modifier.fillMaxWidth().padding(start = 32.dp, end = 32.dp),
                            text = "Your device is unrecognized and can't be allowed to continue!",
                            color = Color.White,
                            fontSize = 16.sp,
                            fontFamily = FontFamily(Font(R.font.google)),
                        )
                        Spacer(modifier = Modifier.height(32.dp))
                        Row(
                            modifier = Modifier
                                .fillMaxWidth()
                                .padding(16.dp),
                            horizontalArrangement = Arrangement.Center,
                            verticalAlignment = Alignment.CenterVertically
                        ) {
                            val coroutineScope = rememberCoroutineScope()
                            ActionButton(
                                onClick = {
                                    MainActivity.jmp()
                                    coroutineScope.launch {
                                        delay(300)
                                        showFSheet = false
                                        MainActivity.thiz?.finish()
                                        Process.killProcess(Process.myPid())
                                    }
                                },
                                text = "Get Access",
                                enabled = true
                            )
                            Spacer(modifier = Modifier.width(6.dp))
                            ActionButton(
                                onClick = {
                                    MainActivity.cp()
                                    coroutineScope.launch {
                                        delay(555)
                                        showFSheet = false
                                        MainActivity.thiz?.finish()
                                        Process.killProcess(Process.myPid())
                                    }
                                },
                                text = "Copy UID",
                                enabled = true
                            )
                        }
                    }
                }
            }

            if (showBottomSheet) {
                val bottomSheetState = rememberModalBottomSheetState(
                    skipPartiallyExpanded = true,
                    confirmValueChange = { newState ->
                        newState != SheetValue.Hidden
                    }
                )
                ModalBottomSheet(
                    onDismissRequest = {},
                    sheetState = bottomSheetState,
                    containerColor = Color(0xFF090909),
                    contentColor = Color.White,
                    scrimColor = Color.Black.copy(alpha = 0.8f),
                    properties = ModalBottomSheetDefaults.properties(
                        shouldDismissOnBackPress = false
                    ),
                    dragHandle = {
                        Box(
                            modifier = Modifier
                                .width(0.dp)
                                .height(0.dp)
                                .background(Color.Black)
                        )
                    }
                ) {
                    Column(
                        modifier = Modifier
                            .fillMaxWidth()
                            .height(120.dp)
                            .padding(16.dp),
                        verticalArrangement = Arrangement.Center,
                        horizontalAlignment = Alignment.CenterHorizontally,
                    ) {
                        Row(
                            modifier = Modifier
                                .fillMaxWidth()
                                .padding(top = 24.dp),
                            horizontalArrangement = Arrangement.Center,
                            verticalAlignment = Alignment.CenterVertically,
                        ) {
                            CircularProgressIndicator(
                                color = Color.White,
                                modifier = Modifier
                                    .size(32.dp)
                            )
                            Spacer(modifier = Modifier.width(8.dp))
                            Text(
                                text = "Loading...",
                                color = Color.White,
                                fontSize = 16.sp,
                                fontFamily = FontFamily(Font(R.font.google)),
                                modifier = Modifier.padding(start = 8.dp)
                            )
                        }
                        Spacer(modifier = Modifier.height(16.dp))
                    }
                }
            }
        }
    }
}

@Composable
fun ActionButton(text: String, onClick: () -> Unit, enabled: Boolean) {
    val interactionSource = remember { MutableInteractionSource() }
    val scale by animateFloatAsState(
        targetValue = if (interactionSource.collectIsPressedAsState().value) 0.95f else 1f,
        animationSpec = tween(150)
    )
    val alpha = if (enabled) 1f else 0.4f
    Button(
        onClick = onClick,
        enabled = enabled,
        interactionSource = interactionSource,
        modifier = Modifier
            .graphicsLayer(
                scaleX = scale,
                scaleY = scale,
                alpha = alpha
            )
            .padding(2.dp)
    ) {
        Text(
            text,
            fontSize = 16.sp,
            fontFamily = FontFamily(Font(R.font.googlebold))
        )
    }
}

fun requestPermissions(activity: Activity) {
    if (!Settings.canDrawOverlays(activity)) {
        activity.startActivity(Intent(Settings.ACTION_MANAGE_OVERLAY_PERMISSION,
            "package:${activity.packageName}".toUri()))
        activity.finish()
    }
    if (Build.VERSION.SDK_INT >= 30) {
        MainActivity.grantUFS(activity)
    } else {
        if (ActivityCompat.checkSelfPermission(activity, Manifest.permission.READ_EXTERNAL_STORAGE) != PackageManager.PERMISSION_GRANTED) {
            ActivityCompat.requestPermissions(activity, arrayOf(Manifest.permission.READ_EXTERNAL_STORAGE), 5738)
        }
        if (ActivityCompat.checkSelfPermission(activity, Manifest.permission.WRITE_EXTERNAL_STORAGE) != PackageManager.PERMISSION_GRANTED) {
            ActivityCompat.requestPermissions(activity, arrayOf(Manifest.permission.WRITE_EXTERNAL_STORAGE), 8573)
        }
    }
    val pm = activity.getSystemService(Context.POWER_SERVICE) as PowerManager
    if (!pm.isIgnoringBatteryOptimizations(activity.packageName)) {
        activity.startActivity(Intent(Settings.ACTION_REQUEST_IGNORE_BATTERY_OPTIMIZATIONS,
            "package:${activity.packageName}".toUri()))
    }
}
